File: C++/src/Intersector1D1D.cpp
#include "Intersector1D1D.hpp"

Intersector1D1D::Intersector1D1D()
{
  type = Intersector1D1D::NoIntersection;
  resultParametricCoordinates.setZero(2);
}

// ***************************************************************************
bool Intersector1D1D::ComputeIntersection()
{
  // parallelism is the wedge product of the 2 tangent vectors, so analyzing it we can say if they are parallel
  double parallelism = matrixTangentVector.determinant();
  type = NoIntersection;
  bool intersection = false;

  // check is used to not use square root function and division, lowering the computational cost. In fact we are controlling if the wedge product
  // divided by the squared norms of the two tangent vectors multiplied is greater than squared parallelism
  double check = INTERSECTION_TOLERANCE * INTERSECTION_TOLERANCE * matrixTangentVector.col(0).squaredNorm() *  matrixTangentVector.col(1).squaredNorm();
  //checking if the vectors are not parallel
  if(parallelism * parallelism >= check)
  {
    Matrix2d solverMatrix;
    //inverting the matrix made up by the 2 tangent vectors multiplied by the determinant
    solverMatrix << matrixTangentVector(1,1), -matrixTangentVector(0,1), -matrixTangentVector(1,0), matrixTangentVector(0,0);
    /// <li> If the edge and the trace are not parallel look for the intersection with parametric coordinates
    resultParametricCoordinates = solverMatrix * rightHandSide;
    //dividing by the determinant here because we are computing only 2 divisions
    resultParametricCoordinates /= parallelism;
    //checking if the parametric coordinates of the second segment is inside the segment itself
    if (resultParametricCoordinates(1) > -INTERSECTION_TOLERANCE  && resultParametricCoordinates(1)-1.0 < INTERSECTION_TOLERANCE)
    {
      //now we can say that there is intersection because the line of the first segment intersect the second segment and thus the intesection
      //is at least "IntersectionOnLine"
      type = IntersectionOnLine;
      intersection = true;
      //checking if the parametric coordinates of the intersection, in respect to the first segment, is inside the segment itself
      if (resultParametricCoordinates(0) > -INTERSECTION_TOLERANCE  && resultParametricCoordinates(0)-1.0 < INTERSECTION_TOLERANCE)
        //if it is true the intersection is inside the segment
        type = IntersectionOnSegment;
    }
  }
  //if they are parallel
  else
  {
    //parallelism2 is the wedge product between the tangent vector of the first segment and the vector made up of (origin of the second segment) - (origin of the first segment),represented by rightHandSide
    //parallelism2 is used to control if the segments are on the same line: if the wedge product is 0 we can say that they are parallel on the same line because
    //rightHandSide will be equal to (0,0), whereas in the other case it will be different from 0
    double parallelism2 = fabs(matrixTangentVector(0,0) * rightHandSide.y() - rightHandSide.x() * matrixTangentVector(1,0));
    /// <li> In case of parallelism check if the segment is the same with parametric coordinates

    double squaredNormFirstEdge = matrixTangentVector.col(0).squaredNorm();
    //check 2 is used just like the previous check
    double check2 = INTERSECTION_TOLERANCE * INTERSECTION_TOLERANCE * squaredNormFirstEdge * rightHandSide.squaredNorm();
    //checking if the segments are on the same line, if not there is "NoIntersection"
    if( parallelism2 * parallelism2 <= check2 )
    {
      //dividing here to reduce computational cost
      double tempNorm = 1.0/squaredNormFirstEdge;
      //inverting the tangent vector to solve the system for the first parametric coordinate
      resultParametricCoordinates(0) = matrixTangentVector.col(0).dot(rightHandSide) * tempNorm ;
      // solving for the second parametric coordinate
      resultParametricCoordinates(1) = resultParametricCoordinates(0) - matrixTangentVector.col(0).dot(matrixTangentVector.col(1)) * tempNorm;
      // we know that there is intersection
      intersection = true;
      type = IntersectionParallelOnLine;

      //swapping in order to have ascending order
      if(resultParametricCoordinates(1) < resultParametricCoordinates(0))
      {
        double tmp = resultParametricCoordinates(0);
        resultParametricCoordinates(0) = resultParametricCoordinates(1);
        resultParametricCoordinates(1) = tmp;
      }
      // if one vertex is in the edge there is the intersection
      if( (resultParametricCoordinates(0) > -INTERSECTION_TOLERANCE && resultParametricCoordinates(0)-1.0 < INTERSECTION_TOLERANCE) ||
          (resultParametricCoordinates(1) > -INTERSECTION_TOLERANCE && resultParametricCoordinates(1)-1.0 <INTERSECTION_TOLERANCE)   )
        type = IntersectionParallelOnSegment;
      else
      {
        //IL PRIMO SEGMENTO DATO IN INPUT E' CONTENUTO NEL SECONDO O VICEVERSA
        if( ( resultParametricCoordinates(0) < INTERSECTION_TOLERANCE && resultParametricCoordinates(1) - 1.0 > -INTERSECTION_TOLERANCE) )
          type = IntersectionParallelOnSegment;
      }
    }
  }
  /// </ul>
  if(resultParametricCoordinates(0) < -INTERSECTION_TOLERANCE || resultParametricCoordinates(0) > 1.0 + INTERSECTION_TOLERANCE)
    positionIntersectionFirstEdge =  Outer;
  else if((resultParametricCoordinates(0) > -INTERSECTION_TOLERANCE) && (resultParametricCoordinates(0) < INTERSECTION_TOLERANCE))
  {
    resultParametricCoordinates(0) = 0.0;
    positionIntersectionFirstEdge = Begin;
  }
  else if ((resultParametricCoordinates(0) > 1.0 - INTERSECTION_TOLERANCE) && (resultParametricCoordinates(0) < 1.0 +INTERSECTION_TOLERANCE))
  {
    resultParametricCoordinates(0) = 1.0;
    positionIntersectionFirstEdge = End;
  }
  else
    positionIntersectionFirstEdge = Inner;

  if(resultParametricCoordinates(1) < -INTERSECTION_TOLERANCE || resultParametricCoordinates(1) > 1.0 + INTERSECTION_TOLERANCE)
    positionIntersectionSecondEdge =  Outer;
  else if((resultParametricCoordinates(1) > -INTERSECTION_TOLERANCE) && (resultParametricCoordinates(1) < INTERSECTION_TOLERANCE))
  {
    resultParametricCoordinates(1) = 0.0;
    positionIntersectionSecondEdge = Begin;
  }
  else if ((resultParametricCoordinates(1) > 1.0 - INTERSECTION_TOLERANCE) && (resultParametricCoordinates(1) < 1.0 + INTERSECTION_TOLERANCE))
  {
    resultParametricCoordinates(1) = 1.0;
    positionIntersectionSecondEdge = End;
  }
  else
    positionIntersectionSecondEdge = Inner;

  return intersection;
}

File: Python/src/Intersector1D1D.py
from src import CONSTANT
import numpy as np
import numpy.linalg as nalg
import math
from enum import Enum


class Type(Enum):
    NoIntersection = 0
    IntersectionOnLine = 1
    IntersectionOnSegment = 2
    IntersectionParallelOnLine = 3
    IntersectionParallelOnSegment = 4


class Position(Enum):
    Begin = 0
    Inner = 1
    End = 2
    Outer = 3


class IIntersector1D1D:
    def set_first_segment(self, origin, end):
        pass

    def set_second_segment(self, origin, end):
        pass

    def compute_intersection(self):
        pass

    def parametric_coordinates(self):
        pass

    def first_parametric_coordinate(self):
        pass

    def second_parametric_coordinate(self):
        pass

    def position_intersection_in_first_segment(self):
        pass

    def position_intersection_in_second_segment(self):
        pass

    def type_intersection(self):
        pass

    def intersection_first_parametric_coordinate(self, origin, end):
        pass

    def intersection_second_parametric_coordinate(self, origin, end):
        pass


class Intersector1D1D(IIntersector1D1D):
    def __init__(self):
        self.__type = Type.NoIntersection
        self.__positionIntersectionFirstEdge = Position.Begin
        self.__positionIntersectionSecondEdge = Position.Begin
        self.__resultParametricCoordinates = np.array([0, 0])
        self.__originFirstSegment = np.array([0, 0])
        self.__rightHandSide = np.array([0, 0])
        self.__matrixTangentVector = np.zeros((2, 2))

    def set_first_segment(self, origin, end):
        self.__matrixTangentVector[:, 0] = end - origin
        self.__originFirstSegment = origin

    def set_second_segment(self, origin, end):
        self.__matrixTangentVector[:, 1] = origin - end
        self.__rightHandSide = origin - self.__originFirstSegment

    def parametric_coordinates(self):
        return self.__resultParametricCoordinates

    def first_parametric_coordinate(self):
        return self.__resultParametricCoordinates[0]

    def second_parametric_coordinate(self):
        return self.__resultParametricCoordinates[1]

    def position_intersection_in_first_segment(self):
        return self.__positionIntersectionFirstEdge

    def position_intersection_in_second_segment(self):
        return self.__positionIntersectionSecondEdge

    def type_intersection(self):
        return self.__type

    def intersection_first_parametric_coordinate(self, origin, end):
        return (1 - self.__resultParametricCoordinates[0]) * origin + self.__resultParametricCoordinates[0] * end

    def intersection_second_parametric_coordinate(self, origin, end):
        return (1 - self.__resultParametricCoordinates[1]) * origin + self.__resultParametricCoordinates[1] * end

    def compute_intersection(self):
        parallelism = nalg.det(self.__matrixTangentVector)
        self.__type = Type.NoIntersection
        intersection = False
        check = CONSTANT.INTERSECTION_TOLERANCE * CONSTANT.INTERSECTION_TOLERANCE * nalg.norm(
            self.__matrixTangentVector[:, 0]) * nalg.norm(self.__matrixTangentVector[:, ])

        if parallelism * parallelism >= check:
            solverMatrix = np.array([[self.__matrixTangentVector[1, 1], -self.__matrixTangentVector[0, 1]],
                                     [-self.__matrixTangentVector[1, 0], self.__matrixTangentVector[0, 0]]])

            self.__resultParametricCoordinates = np.dot(solverMatrix, self.__rightHandSide)
            self.__resultParametricCoordinates /= parallelism

            if (self.__resultParametricCoordinates[1] > -CONSTANT.INTERSECTION_TOLERANCE and
                    self.__resultParametricCoordinates[1] - 1.0 < CONSTANT.INTERSECTION_TOLERANCE):
                self.__type = Type.IntersectionOnLine
                intersection = True

                if (self.__resultParametricCoordinates[0] > -CONSTANT.INTERSECTION_TOLERANCE and
                        self.__resultParametricCoordinates[0] - 1.0 < CONSTANT.INTERSECTION_TOLERANCE):
                    self.__type = Type.IntersectionOnSegment

        else:
            parallelism2 = math.fabs(
                self.__matrixTangentVector[0, 0] * self.__rightHandSide[1] - self.__rightHandSide[0] *
                self.__matrixTangentVector[1, 0])
            squaredNormFirstEdge = np.square(nalg.norm(self.__matrixTangentVector[:, 0]))
            check2 = CONSTANT.INTERSECTION_TOLERANCE * CONSTANT.INTERSECTION_TOLERANCE * squaredNormFirstEdge * nalg.norm(
                self.__rightHandSide)
            if (parallelism2 * parallelism2 <= check2):
                tempNorm = 1.0 / squaredNormFirstEdge
                self.__resultParametricCoordinates[0] = np.dot(self.__matrixTangentVector[:, 0], self.__rightHandSide)
                self.__resultParametricCoordinates[1] = self.__resultParametricCoordinates[0] - np.dot(
                    self.__matrixTangentVector[:, 0], self.__matrixTangentVector[:, 1])
                self.__resultParametricCoordinates = tempNorm * self.__resultParametricCoordinates
                intersection = True
                self.__type = Type.IntersectionParallelOnLine

                if (self.__resultParametricCoordinates[1] < self.__resultParametricCoordinates[0]):
                    tmp = self.__resultParametricCoordinates[0]
                    self.__resultParametricCoordinates[0] = self.__resultParametricCoordinates[1]
                    self.__resultParametricCoordinates[1] = tmp

                if ((self.__resultParametricCoordinates[0] > -CONSTANT.INTERSECTION_TOLERANCE and
                     self.__resultParametricCoordinates[0] - 1.0 < CONSTANT.INTERSECTION_TOLERANCE) or (
                        self.__resultParametricCoordinates[1] > -CONSTANT.INTERSECTION_TOLERANCE and
                        self.__resultParametricCoordinates[1] - 1.0 < CONSTANT.INTERSECTION_TOLERANCE)):
                    self.__type = Type.IntersectionParallelOnSegment

                else:
                    if (self.__resultParametricCoordinates[0] < CONSTANT.INTERSECTION_TOLERANCE and
                            self.__resultParametricCoordinates[
                                1] - 1.0 > -CONSTANT.INTERSECTION_TOLERANCE):
                        self.__type = Type.IntersectionParallelOnSegment

        if (self.__resultParametricCoordinates[0] < -CONSTANT.INTERSECTION_TOLERANCE or
                self.__resultParametricCoordinates[0] - 1.0 > CONSTANT.INTERSECTION_TOLERANCE):
            self.__positionIntersectionFirstEdge = Position.Outer

        elif ((self.__resultParametricCoordinates[0] > -CONSTANT.INTERSECTION_TOLERANCE) and (
                self.__resultParametricCoordinates[0] < CONSTANT.INTERSECTION_TOLERANCE)):
            self.__resultParametricCoordinates[0] = 0.0
            self.__positionIntersectionFirstEdge = Position.Begin

        elif ((self.__resultParametricCoordinates[0] > 1.0 - CONSTANT.INTERSECTION_TOLERANCE) and (
                self.__resultParametricCoordinates[0] < 1.0 + CONSTANT.INTERSECTION_TOLERANCE)):
            self.__resultParametricCoordinates[0] = 1.0
            self.__positionIntersectionFirstEdge = Position.End

        else:
            self.__positionIntersectionFirstEdge = Position.Inner

        if (self.__resultParametricCoordinates[1] < -CONSTANT.INTERSECTION_TOLERANCE or
                self.__resultParametricCoordinates[1] > 1.0 + CONSTANT.INTERSECTION_TOLERANCE):
            self.__positionIntersectionSecondEdge = Position.Outer

        elif ((self.__resultParametricCoordinates[1] > -CONSTANT.INTERSECTION_TOLERANCE) and (
                self.__resultParametricCoordinates[1] < CONSTANT.INTERSECTION_TOLERANCE)):
            self.__resultParametricCoordinates[1] = 0.0
            self.__positionIntersectionSecondEdge = Position.Begin

        elif ((self.__resultParametricCoordinates[1] > 1.0 - CONSTANT.INTERSECTION_TOLERANCE) and (
                self.__resultParametricCoordinates[1] < 1.0 + CONSTANT.INTERSECTION_TOLERANCE)):
            self.__resultParametricCoordinates[1] = 1.0
            self.__positionIntersectionSecondEdge = Position.End

        else:
            self.__positionIntersectionSecondEdge = Position.Inner

        return intersection
